contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+1. Admin statically creates PVs and/or StorageClasses.
+2. User creates unbound PVC and there are no prebound PVs for it.
+3. **NEW:** PVC binding and provisioning is delayed until a pod is created that
+references it.
+4. User creates a pod that uses the PVC.
+5. Pod starts to get processed by the scheduler.
+6. **NEW:** A new predicate function, called MatchUnboundPVCs, will look at all of
+a Pod’s unbound PVCs, and try to find matching PVs for that node based on the
+PV topology.  If there are no matching PVs, then it checks if dynamic
+provisioning is possible for that node.
+7. **NEW:** The scheduler continues to evaluate priorities.  A new priority
+function, called PrioritizeUnboundPVCs, will get the PV matches per PVC per
+node, and compute a priority score based on various factors.
+8. **NEW:** After evaluating all the existing predicates and priorities, the
+scheduler will pick a node, and call a new assume function, AssumePVCs,
+passing in the Node.  The assume function will check if any binding or
+provisioning operations need to be done.  If so, it will update the PV cache to
+mark the PVs with the chosen PVCs.
+9. **NEW:** If PVC binding or provisioning is required, we do NOT AssumePod.
+Instead, a new bind function, BindPVCs, will be called asynchronously, passing
+in the selected node.  The bind function will prebind the PV to the PVC, or
+trigger dynamic provisioning, and then wait until the PVCs are bound
+successfully or encounter failure.  Then, it always sends the Pod through the
+scheduler again for reasons explained later.
+10. When a Pod makes a successful scheduler pass once all PVCs are bound, the
+scheduler assumes and binds the Pod to a Node.
+11. Kubelet starts the Pod.
+
+This new workflow will have the scheduler handle unbound PVCs by choosing PVs
+and prebinding them to the PVCs.  The PV controller completes the binding
+transaction, handling it as a prebound PV scenario.
+
+Prebound PVCs and PVs will still immediately be bound by the PV controller.
+
+One important point to note is that for the alpha phase, manual recovery is
+required in following error conditions:
+* A Pod has multiple PVCs, and only a subset of them successfully bind.
+* The scheduler chose a PV and prebound it, but the PVC could not be bound.
+
+In both of these scenarios, the primary cause for these errors is if a user
+prebinds other PVCs to a PV that the scheduler chose.  Some workarounds to
+avoid these error conditions are to:
+* Prebind the PV instead.
+* Separate out volumes that the user prebinds from the volumes that are
+available for the system to choose from by StorageClass.
+
+In addition, this new design will cause functional regression for the
+following use cases:
+* While a Pod is running, a user deletes the PVC and PV, and then recreates them.
+    * With the current behavior, the PVC will rebind to the same or different PV.
+    * With this new behavior, the PVC will not rebind because binding would only
```

Comment:
On EBS at least, since you can't delete a volume that is mounted and in use. Deleting a dynamically bound PVC results in PV being "Released" (I think).  So when user creates a new PVC with same name (because user does not want to change pod's yaml), a new PV is created. So we get 2 PVs (one released and one bound) mapped to same PVC.

Reply:
Not all storage providers may be able to detect if a volume is in use before deleting. But assuming it can detect and fail the deletion, let's break this up into 2 scenarios actually: 

1. Delete PVC, PV is released, but not deleted.  Recreate PVC.
If the container restarts, the old volume is still used in the new container.  If the Pod restarts, then it goes back through scheduling, and we rebind the PVC at that time.  Pod termination works because we only need the PV, not the PVC object still exists.  So no regression here.

2. Delete both PVC and PV objects.  Recreate PVC and PV objects.
Pod termination still works because PV object is still in the actual state.  If kubelet restarts, volume reconstruction will recreate the PV object.  No regression here.

So actually, I made a wrong assumption that we needed the PVC object to be bound for Pod teardown.  kubelet volume manager only needs the PV object, and it can reconstruct it too if kubelet crashes.  Of course, this also assumes that the volume plugin implements reconstruction correctly.

I'll remove this section unless we can think of another scenario that won't work with the new design.  I'll leave the discussion of if we want to prevent PVC and PV deletion when it's still in use to a separate issue.